### PR TITLE
Fix xamarin-android issue 1130 and add test

### PR DESCRIPTION
**Pull request by kg**
https://github.com/xamarin/xamarin-android/issues/1130 is caused by us not properly storing/restoring the Thread.CurrentPrincipal context when execution contexts are stored. Task (and I think ThreadPool?) store the current context when a job is queued and restore it when running the job, which carries over state like the thread's current principal.

Right now we have a weird hand-rolled implementation of CurrentPrincipal that I don't quite understand. I think we should be using the CurrentPrincipal value stored in the BCL's CallContext, so I tried doing that and it seems to fix this without breaking anything.

I'd appreciate getting suggestions on how we should go about doing this for real. Does anyone know why the old implementation works the way it does? I assume the serialization logic has something to do with appdomains, but I would not have expected principals for the current thread to be shared between domains.

### CODE REVIEW

**Comment by kg** [comment]
@monojenkins build Linux i386

**Comment by kg** [comment]
@monojenkins build Windows x64

**Comment by dazinator** [comment]
@kg Thanks for looking at this. 
I am not familiar with the mono implementation, but my understanding is that a Principal needs to be serialiazable as that is how it is copied to new threads. Each thread should have its own instance of the Principal and not share an instance. I believe this is so that if one thread mutates its CurrentPrincipal, it doesn't affect any other currently running threads - as they all have their own instance.
Full.NET had something called "SecurityContext" which it stored in the ExecutionContext and it would restore this SecurityContext onto any new threads as part of `ExecutionContext.Run`. This would serialize a copy of the captured Principal and use that to set the Principal on the new thread.

File: mcs/class/System.Core/System.Core_test.dll.sources [hunk 1]
```diff
@@ -82,3 +82,4 @@ System.Runtime.CompilerServices/DynamicAttributeTest.cs
 System.Linq/ParallelEnumerableTests.cs
 System.Linq/ParallelTestHelper.cs
 System.Threading.Tasks/TaskExtensionsTests.cs
+System.Threading.Tasks/ThreadPrincipalTests.cs
```
Comment (marek-safar) [inline mcs/class/System.Core/System.Core_test.dll.sources]: Why is this test here and not in mscorlib ?

Comment (kg) [reply mcs/class/System.Core/System.Core_test.dll.sources]: Added the reporter's test before I identified the fix. I can move it.

Comment (luhenry) [reply mcs/class/System.Core/System.Core_test.dll.sources]: This should definitely be in `mscorlib` test suite

**Comment by kg** [comment]
@dazinator Thanks, that makes sense. So an assumption here is that any time a principal is transferred to another thread, it's copied. Do you happen to know any more about any other assumptions your code might rely on, like is the same copy kept around threadlocal style? Serializing/deserializing principals every time code moves to a new thread seems like it would be expensive.

**Comment by dazinator** [comment]
I believe the principal flows with ExecutionContext.Run (i.e whenever wanting to transfer to another thread / synchronisationcontext) so I think its necessary to copy the principal at that point. Are you saying perhaps the new thread is being reused and already has a copy of the principal from a previous execution? If so I dont think you can avoid a copy still on each run still, because the principal on the parent thread may change between runs. I.e

1. Set principal on main thread 
2. Async code runs on thread 2 - principal is copied from main thread.
3. Set new principal on main thread
4. Async code runs on thread 2 - principal should be copied again from main thread

At step 4 above, even though thread 2 could already have a principal set, it nreds to get set again because it needs to match the context from main / parent thread.

If you know of some way to optimise this check, perhaps on mono the principal could optionally derive some interface so that we can only perform the copy if it doesnt match? That would be good. I am assuming AsyncLocal can be ruled out as it doesn't copy instances but shares them accross threads.

To be honest, sharing the same instance of a principal accross threads seems like a framework decision that probably should explore a bit. It avoids the serialisation issue but what would pitfalls be?

1. I don't think you'd ever want to mutate an existing instance of a principal anyway, but rather, you would set a new instance into Thread.CurrentPrincipal - i.e typically you'd treat it as immutable. One time you might want to mutate the principal / identity is to add roles at runtime.

2. If you did mutate the instance (i.e lets say add a role to its "roles" collection) - if people knew that mono did not copy instances but shared them accross threads, then there custom pricipals would need to be written to be thread safe. I.e the roles collection would need to be a concurrent collection for example.

I think 2. might be an issue if the existing framework principals (claimsprincipal, genericprincipal) are not thread safe and allow you to mutate them. I.e you could add a claim on one thread whilst another thread is checking the claims..

So given the above, it probably is better to stick with the net framework design that principals are copied and not shared accross threads. But how you can optimise that to avoid redundant serialisations im not sure. I'd be happy to derive additional interfaces on my principal / identity classes to support such an optimisation check if mono added one - if that helps at all! I'd be surprised if the .net framework does any optimisation around this.

**Comment by dazinator** [comment]
@kg 
>  Do you happen to know any more about any other assumptions your code might rely on, like is the same copy kept around threadlocal style? Serializing/deserializing principals every time code moves to a new thread seems like it would be expensive

Just occurred to me, do you mean, keep a copy around on the parent thread - ready to set on each new thread so it doesnt need to be serialised "on demand"?
If so, problem I see with that:

1. You would have to keep a copy, per thread. Otherwise they would share same instance.
2. If a child thread mutates that copy, it woukd need to be restored again after execution..

**Comment by dazinator** [comment]
@kg 
I think the only time you can avoid a copy is when all of the following are true:

Prior to execution on the new thread:
1. Principal / Identity of parent thread has already been copied to new thread in the past.
2. Principal / Identity of parent thread has not changed since that copy occurred.
3. Principal / Identity of new thread has not changed since that copy occurred.
 In otherwords.. they are still the same! :-)

**Comment by kg** [comment]
I did some research based on @dazinator 's info and I think our current test coverage for this functionality is insufficient, so I need to add more tests before I can merge this.

**Comment by dazinator** [comment]
@kg can I be of any further assistance?

**Comment by luhenry** [comment]
@monojenkins build

**Comment by kg** [comment]
I haven't had time to go back to this due to other work priorities.

**Comment by dazinator** [comment]
@luhenry - thank you that would be great. I'd love to see this in a future release of mono, it would allow me to remove my hacky workarounds :-)

**Comment by luhenry** [comment]
`Linux ARMv7 hard float` and `Linux ARMv5 soft float` are to be ignored as they are status check that ran more than a month ago; they are still here because of the rename from `Linux ARMv5 soft float` to `Linux ARMv5`, and same for `Linux ARMv7`.

**Comment by dazinator** [comment]
Thanks all involved for your work in resolving this.

**Pull request merged**